File: tools/pcd/structural_map.py
"""PCD Round 0a: Deterministic structural map generator.

Produces dependency graphs and file clusters from import analysis.
Uses CAFI for file discovery and optional description enrichment.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone

# Add the tools directory to sys.path so we can import shared _lib and pcd packages.
_pcd_dir = os.path.dirname(os.path.abspath(__file__))
_tools_dir = os.path.dirname(_pcd_dir)
if _tools_dir not in sys.path:
    sys.path.insert(0, _tools_dir)

from _lib.cache import load_cache  # noqa: E402
from _lib.config import load_config  # noqa: E402
from _lib.file_walker import walk_project  # noqa: E402
from pcd.import_parser import (  # noqa: E402
    ImportEntry,
    detect_language,
    parse_imports,
)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

OUTPUT_DIR = os.path.join(".prove", "steward", "pcd")
OUTPUT_FILE = "structural-map.json"
CACHE_PATH = os.path.join(".prove", "file-index.json")


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _count_lines(file_path: str) -> int:
    """Count lines in a file. Return 0 on error."""
    try:
        with open(file_path, "r", encoding="utf-8", errors="replace") as f:
            return sum(1 for _ in f)
    except OSError:
        return 0


def _resolve_import_to_file(
    module: str,
    language: str,
    project_files: set[str],
    project_root: str,
) -> str | None:
    """Resolve a module name to a project-relative file path.

    Parameters
    ----------
    module:
        The imported module string (e.g. ``cafi.hasher``, ``./utils``).
    language:
        Language of the importing file.
    project_files:
        Set of known project-relative file paths.
    project_root:
        Absolute path to the project root.

    Returns
    -------
    The matched project-relative file path, or ``None`` if unresolvable.
    """
    if language == "python":
        return _resolve_python(module, project_files)
    if language == "rust":
        return _resolve_rust(module, project_files)
    if language == "go":
        return _resolve_go(module, project_files, project_root)
    if language in ("javascript", "typescript"):
        return _resolve_js_ts(module, project_files)
    return None


def _resolve_python(module: str, project_files: set[str]) -> str | None:
    """Resolve a Python module path to a file.

    ``cafi.hasher`` -> try ``cafi/hasher.py``, ``cafi/hasher/__init__.py``
    Relative imports (starting with ``.``) are not resolved here since
    we lack the importing file's package context in this simple resolver.
    """
    if module.startswith("."):
        return None
    parts = module.split(".")
    # Try as a direct module file: a/b/c.py
    candidate = os.path.join(*parts) + ".py" if parts else None
    if candidate and candidate in project_files:
        return candidate
    # Try as a package: a/b/c/__init__.py
    candidate_pkg = os.path.join(*parts, "__init__.py") if parts else None
    if candidate_pkg and candidate_pkg in project_files:
        return candidate_pkg
    return None


def _resolve_rust(module: str, project_files: set[str]) -> str | None:
    """Resolve a Rust use path to a file.

    ``crate::parser::Parser`` -> try ``src/parser.rs``, ``src/parser/mod.rs``
    """
    if not module.startswith(("crate::", "self::", "super::")):
        return None
    # Strip the crate/self/super prefix
    parts = module.split("::")
    # Remove the first segment (crate/self/super)
    path_parts = parts[1:]
    if not path_parts:
        return None
    # For crate:: imports, the root is typically src/
    prefix = "src" if parts[0] == "crate" else ""
    # Drop the last part if it looks like a type name (starts with uppercase)
    if path_parts and path_parts[-1] and path_parts[-1][0].isupper():
        path_parts = path_parts[:-1]
    if not path_parts:
        return None
    if prefix:
        base = os.path.join(prefix, *path_parts)
    else:
        base = os.path.join(*path_parts)
    # Try as file.rs
    candidate = base + ".rs"
    if candidate in project_files:
        return candidate
    # Try as directory/mod.rs
    candidate_mod = os.path.join(base, "mod.rs")
    if candidate_mod in project_files:
        return candidate_mod
    return None


def _resolve_go(
    module: str, project_files: set[str], project_root: str
) -> str | None:
    """Resolve a Go import path to a directory with .go files.

    Go imports are package paths. We try to strip the module prefix
    (from go.mod) and look for .go files in the resulting directory.
    Returns the first matching .go file path for the adjacency graph.
    """
    # Read go.mod to find the module prefix
    go_mod_path = os.path.join(project_root, "go.mod")
    module_prefix = ""
    try:
        with open(go_mod_path, "r") as f:
            for line in f:
                line = line.strip()
                if line.startswith("module "):
                    module_prefix = line.split(None, 1)[1].strip()
                    break
    except OSError:
        pass

    if module_prefix and module.startswith(module_prefix):
        rel = module[len(module_prefix) :].lstrip("/")
    else:
        rel = module

    # Look for .go files in that directory
    for pf in sorted(project_files):
        if pf.startswith(rel + "/") and pf.endswith(".go"):
            return pf
        # Also try exact match for single-file packages
        if pf == rel + ".go":
            return pf
    return None


def _resolve_js_ts(module: str, project_files: set[str]) -> str | None:
    """Resolve a JS/TS relative import to a file.

    ``./utils`` -> try ``utils.js``, ``utils.ts``, ``utils.tsx``,
    ``utils/index.js``, ``utils/index.ts``, etc.
    """
    if not module.startswith(("./", "../")):
        return None
    # Normalize the path (remove leading ./)
    base = os.path.normpath(module)
    extensions = [".js", ".jsx", ".ts", ".tsx", ".mjs"]
    # If already has an extension, try directly
    _, ext = os.path.splitext(base)
    if ext in extensions:
        if base in project_files:
            return base
        return None
    # Try with each extension
    for e in extensions:
        candidate = base + e
        if candidate in project_files:
            return candidate
    # Try as directory with index file
    for e in extensions:
        candidate = os.path.join(base, "index" + e)
        if candidate in project_files:
            return candidate
    return None


# ---------------------------------------------------------------------------
# Core pipeline functions
# ---------------------------------------------------------------------------


def _build_dependency_graph(
    files: list[str], project_root: str
) -> tuple[list[ImportEntry], dict[str, list[str]]]:
    """Build a dependency graph from parsed imports.

    Parameters
    ----------
    files:
        List of project-relative file paths to analyze.
    project_root:
        Absolute path to the project root.

    Returns
    -------
    Tuple of (all_imports, adjacency) where adjacency maps each file
    to the list of internal files it imports from.
    """
    project_files = set(files)
    all_imports: list[ImportEntry] = []
    adjacency: dict[str, list[str]] = {f: [] for f in files}

    for rel_path in files:
        full_path = os.path.join(project_root, rel_path)
        try:
            with open(full_path, "r", encoding="utf-8", errors="replace") as fh:
                content = fh.read()
        except OSError:
            continue

        imports = parse_imports(rel_path, content)
        all_imports.extend(imports)

        language = detect_language(rel_path)
        if language is None:
            continue

        seen_targets: set[str] = set()
        for imp in imports:
            if imp.import_type not in ("internal", "external"):
                continue
            target = _resolve_import_to_file(
                imp.imported_module, language, project_files, project_root
            )
            if target and target != rel_path and target not in seen_targets:
                seen_targets.add(target)
                adjacency[rel_path].append(target)

    return all_imports, adjacency


def _cluster_files(
    files: list[str],
    adjacency: dict[str, list[str]],
    max_cluster_size: int = 15,
) -> list[dict]:
    """Cluster files by dependency connectivity and directory proximity.

    1. Find connected components in the undirected dependency graph.
    2. Split components larger than ``max_cluster_size`` by directory subtree.
    3. Isolated files (no edges) are grouped by their parent directory.

    Parameters
    ----------
    files:
        All file paths to cluster.
    adjacency:
        Directed adjacency list (file -> list of targets).
    max_cluster_size:
        Maximum files per cluster before splitting.

    Returns
    -------
    List of cluster dicts with keys: id, name, files, internal_edges, external_edges.
    """
    # Build undirected adjacency for component finding
    undirected: dict[str, set[str]] = {f: set() for f in files}
    all_edges: set[tuple[str, str]] = set()
    for src, targets in adjacency.items():
        for tgt in targets:
            if tgt in undirected:
                undirected[src].add(tgt)
                undirected[tgt].add(src)
                all_edges.add((src, tgt))

    # Find connected components via BFS
    visited: set[str] = set()
    components: list[list[str]] = []
    for f in sorted(files):
        if f in visited:
            continue
        component: list[str] = []
        queue = [f]
        while queue:
            node = queue.pop(0)
            if node in visited:
                continue
            visited.add(node)
            component.append(node)
            for neighbor in sorted(undirected.get(node, set())):
                if neighbor not in visited:
                    queue.append(neighbor)
        if component:
            components.append(sorted(component))

    # Split large components by directory subtree
    final_groups: list[list[str]] = []
    for comp in components:
        if len(comp) <= max_cluster_size:
            final_groups.append(comp)
        else:
            # Group by directory
            dir_groups: dict[str, list[str]] = {}
            for f in comp:
                d = os.path.dirname(f) or "."
                dir_groups.setdefault(d, []).append(f)
            for d in sorted(dir_groups):
                group = dir_groups[d]
                # Further split if still too large
                for i in range(0, len(group), max_cluster_size):
                    final_groups.append(group[i : i + max_cluster_size])

    # Build cluster dicts
    clusters: list[dict] = []
    file_to_cluster: dict[str, int] = {}

    for idx, group in enumerate(final_groups):
        cluster_id = idx
        for f in group:
            file_to_cluster[f] = cluster_id

        # Compute common prefix for name
        if len(group) == 1:
            name = os.path.dirname(group[0]) or os.path.splitext(group[0])[0]
        else:
            name = os.path.commonpath(group) if group else "root"
        if not name or name == ".":
            name = "root"

        # Count edges
        internal_edges = 0
        external_edges = 0
        cluster_set = set(group)
        for f in group:
            for tgt in adjacency.get(f, []):
                if tgt in cluster_set:
                    internal_edges += 1
                else:
                    external_edges += 1

        clusters.append(
            {
                "id": cluster_id,
                "name": name,
                "files": group,
                "internal_edges": internal_edges,
                "external_edges": external_edges,
            }
        )

    return clusters


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------


def generate_structural_map(
    project_root: str,
    scope: list[str] | None = None,
) -> dict:
    """Generate a PCD structural map for the project.

    Parameters
    ----------
    project_root:
        Absolute path to the project root.
    scope:
        If provided, only process these files (project-relative paths).
        Used for scoped steward-review mode.

    Returns
    -------
    A dict conforming to ``STRUCTURAL_MAP_SCHEMA``.
    The map is also written to ``.prove/steward/pcd/structural-map.json``.
    """
    project_root = os.path.abspath(project_root)

    if scope is not None:
        files = list(scope)
    else:
        config = load_config(project_root, require=False)
        files = walk_project(
            project_root,
            excludes=config.get("excludes", []),
            max_file_size=config.get("max_file_size", 102400),
        )

    # Build dependency graph
    all_imports, adjacency = _build_dependency_graph(files, project_root)

    # Cluster files
    clusters = _cluster_files(files, adjacency)

    # Build file-to-cluster mapping
    file_to_cluster: dict[str, int] = {}
    for cluster in clusters:
        for f in cluster["files"]:
            file_to_cluster[f] = cluster["id"]

    # Build reverse adjacency (imported_by)
    imported_by: dict[str, list[str]] = {f: [] for f in files}
    for src, targets in adjacency.items():
        for tgt in targets:
            if tgt in imported_by:
                imported_by[tgt].append(src)
    # Deduplicate and sort
    for key in imported_by:
        imported_by[key] = sorted(set(imported_by[key]))

    # Load CAFI cache for descriptions
    cache_path = os.path.join(project_root, CACHE_PATH)
    cafi_cache = load_cache(cache_path)
    cafi_files = cafi_cache.get("files", {})

    # Count lines and build language stats
    total_lines = 0
    language_counts: dict[str, int] = {}

    modules: list[dict] = []
    dependency_edges: list[dict] = []

    for rel_path in sorted(files):
        full_path = os.path.join(project_root, rel_path)
        lines = _count_lines(full_path)
        total_lines += lines

        language = detect_language(rel_path) or "unknown"
        language_counts[language] = language_counts.get(language, 0) + 1

        # CAFI description
        cafi_entry = cafi_files.get(rel_path)
        cafi_desc = cafi_entry.get("description") if cafi_entry else None
        if cafi_desc == "":
            cafi_desc = None

        module_dict: dict = {
            "path": rel_path,
            "lines": lines,
            "language": language,
            "exports": [],  # Export extraction is a future enhancement
            "imports_from": sorted(set(adjacency.get(rel_path, []))),
            "imported_by": imported_by.get(rel_path, []),
            "cluster_id": file_to_cluster.get(rel_path, 0),
        }
        if cafi_desc is not None:
            module_dict["cafi_description"] = cafi_desc

        modules.append(module_dict)

        # Dependency edges
        for tgt in adjacency.get(rel_path, []):
            dependency_edges.append(
                {"from": rel_path, "to": tgt, "type": "internal"}
            )

    structural_map: dict = {
        "version": 1,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "generated_by": "deterministic",
        "summary": {
            "total_files": len(files),
            "total_lines": total_lines,
            "languages": language_counts,
        },
        "modules": modules,
        "clusters": clusters,
        "dependency_edges": dependency_edges,
    }

    # Write output
    output_dir = os.path.join(project_root, OUTPUT_DIR)
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, OUTPUT_FILE)
    with open(output_path, "w") as f:
        json.dump(structural_map, f, indent=2)
        f.write("\n")

    return structural_map

File: packages/cli/src/topics/pcd/structural-map.ts
/**
 * PCD Round 0a: deterministic structural map generator.
 *
 * Ported 1:1 from `tools/pcd/structural_map.py`. Produces dependency graphs
 * and file clusters from import analysis. Reuses the shared project walker
 * and CAFI cache loader so on-disk output stays byte-identical to Python —
 * key order, module traversal order, cluster IDs, and edge order all match.
 */

import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, extname, join, normalize, resolve, sep } from 'node:path';
import { loadCache, loadToolConfig, walkProject } from '@claude-prove/shared';
import { type ImportEntry, type Language, detectLanguage, parseImports } from './import-parser';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const OUTPUT_DIR = join('.prove', 'steward', 'pcd');
export const OUTPUT_FILE = 'structural-map.json';
export const CACHE_PATH = join('.prove', 'file-index.json');

const DEFAULT_MAX_CLUSTER_SIZE = 15;

// ---------------------------------------------------------------------------
// Public output shape
// ---------------------------------------------------------------------------

export interface StructuralMapModule {
  path: string;
  lines: number;
  language: string;
  exports: string[];
  imports_from: string[];
  imported_by: string[];
  cluster_id: number;
  /** Present only when `.prove/file-index.json` has a non-empty description. */
  cafi_description?: string;
}

export interface StructuralMapCluster {
  id: number;
  name: string;
  files: string[];
  internal_edges: number;
  external_edges: number;
}

export interface StructuralMapEdge {
  from: string;
  to: string;
  type: 'internal';
}

export interface StructuralMapSummary {
  total_files: number;
  total_lines: number;
  languages: Record<string, number>;
}

export interface StructuralMap {
  version: 1;
  timestamp: string;
  generated_by: 'deterministic';
  summary: StructuralMapSummary;
  modules: StructuralMapModule[];
  clusters: StructuralMapCluster[];
  dependency_edges: StructuralMapEdge[];
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Count lines in a file. Returns 0 on any I/O error. */
export function _countLines(filePath: string): number {
  let content: string;
  try {
    content = readFileSync(filePath, 'utf8');
  } catch {
    return 0;
  }
  if (content.length === 0) return 0;
  // Python's `sum(1 for _ in f)`: one line per `\n`, plus a final partial
  // line if the file doesn't end with `\n`. Equivalently: number of `\n`s
  // plus 1 if the last char isn't `\n`.
  let count = 0;
  for (let i = 0; i < content.length; i++) {
    if (content[i] === '\n') count++;
  }
  if (content[content.length - 1] !== '\n') count++;
  return count;
}

/**
 * Resolve a module name to a project-relative file path per language rules.
 *
 * Mirrors `_resolve_import_to_file` in the Python source: Python =
 * package-relative dotted paths, Rust = crate-relative `::` paths, Go =
 * module-prefix-relative paths, JS/TS = `./` + `../` relative file specs.
 */
export function _resolveImportToFile(
  module: string,
  language: string,
  projectFiles: Set<string>,
  projectRoot: string,
): string | null {
  if (language === 'python') return resolvePython(module, projectFiles);
  if (language === 'rust') return resolveRust(module, projectFiles);
  if (language === 'go') return resolveGo(module, projectFiles, projectRoot);
  if (language === 'javascript' || language === 'typescript') {
    return resolveJsTs(module, projectFiles);
  }
  return null;
}

function resolvePython(module: string, projectFiles: Set<string>): string | null {
  if (module.startsWith('.')) return null;
  const parts = module.split('.');
  if (parts.length === 0) return null;
  const candidate = `${parts.join(sep)}.py`;
  if (projectFiles.has(candidate)) return candidate;
  const candidatePkg = `${parts.join(sep)}${sep}__init__.py`;
  if (projectFiles.has(candidatePkg)) return candidatePkg;
  return null;
}

function resolveRust(module: string, projectFiles: Set<string>): string | null {
  if (
    !module.startsWith('crate::') &&
    !module.startsWith('self::') &&
    !module.startsWith('super::')
  ) {
    return null;
  }
  const parts = module.split('::');
  const pathParts = parts.slice(1);
  if (pathParts.length === 0) return null;
  const prefix = parts[0] === 'crate' ? 'src' : '';
  // Drop the trailing segment if it looks like a type name (starts uppercase).
  // Mirrors Python's `path_parts[-1][0].isupper()` — only ASCII uppercase A-Z
  // counts since `.isupper()` returns False on digits/punct.
  const last = pathParts[pathParts.length - 1] ?? '';
  const trimmed = last.length > 0 && /^[A-Z]/.test(last) ? pathParts.slice(0, -1) : pathParts;
  if (trimmed.length === 0) return null;
  const base = prefix ? join(prefix, ...trimmed) : join(...trimmed);
  const candidate = `${base}.rs`;
  if (projectFiles.has(candidate)) return candidate;
  const candidateMod = join(base, 'mod.rs');
  if (projectFiles.has(candidateMod)) return candidateMod;
  return null;
}

function resolveGo(module: string, projectFiles: Set<string>, projectRoot: string): string | null {
  // Read go.mod to find the module prefix, if any.
  let modulePrefix = '';
  try {
    const goMod = readFileSync(join(projectRoot, 'go.mod'), 'utf8');
    for (const rawLine of goMod.split('\n')) {
      const line = rawLine.trim();
      if (line.startsWith('module ')) {
        const remainder = line.slice('module '.length).trim();
        modulePrefix = remainder;
        break;
      }
    }
  } catch {
    // No go.mod — fall through with empty prefix.
  }

  let rel: string;
  if (modulePrefix !== '' && module.startsWith(modulePrefix)) {
    rel = module.slice(modulePrefix.length).replace(/^\/+/, '');
  } else {
    rel = module;
  }

  // Match Python's `for pf in sorted(project_files)` order.
  const sorted = [...projectFiles].sort();
  for (const pf of sorted) {
    if (pf.startsWith(`${rel}/`) && pf.endsWith('.go')) return pf;
    if (pf === `${rel}.go`) return pf;
  }
  return null;
}

function resolveJsTs(module: string, projectFiles: Set<string>): string | null {
  if (!module.startsWith('./') && !module.startsWith('../')) return null;
  const base = normalize(module);
  const extensions = ['.js', '.jsx', '.ts', '.tsx', '.mjs'];
  const ext = extname(base);
  if (extensions.includes(ext)) {
    return projectFiles.has(base) ? base : null;
  }
  for (const e of extensions) {
    const candidate = base + e;
    if (projectFiles.has(candidate)) return candidate;
  }
  for (const e of extensions) {
    const candidate = join(base, `index${e}`);
    if (projectFiles.has(candidate)) return candidate;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Core pipeline
// ---------------------------------------------------------------------------

interface DependencyGraph {
  allImports: ImportEntry[];
  adjacency: Map<string, string[]>;
}

/**
 * Parse every file's imports and build a per-source adjacency list of
 * internal + external imports that resolved to another project file.
 * Dedup is per-source in first-appearance order.
 */
export function _buildDependencyGraph(files: string[], projectRoot: string): DependencyGraph {
  const projectFiles = new Set(files);
  const allImports: ImportEntry[] = [];
  const adjacency = new Map<string, string[]>();
  for (const f of files) adjacency.set(f, []);

  for (const relPath of files) {
    const fullPath = join(projectRoot, relPath);
    let content: string;
    try {
      content = readFileSync(fullPath, 'utf8');
    } catch {
      continue;
    }

    const imports = parseImports(relPath, content);
    for (const imp of imports) allImports.push(imp);

    const language = detectLanguage(relPath);
    if (language === null) continue;

    const seen = new Set<string>();
    const targets = adjacency.get(relPath) ?? [];
    for (const imp of imports) {
      if (imp.import_type !== 'internal' && imp.import_type !== 'external') continue;
      const target = _resolveImportToFile(
        imp.imported_module,
        language satisfies Language,
        projectFiles,
        projectRoot,
      );
      if (target !== null && target !== relPath && !seen.has(target)) {
        seen.add(target);
        targets.push(target);
      }
    }
  }

  return { allImports, adjacency };
}

/**
 * Cluster files by dependency connectivity then by directory proximity.
 *
 * Algorithm (byte-matched to Python's `_cluster_files`):
 *   1. BFS connected components on the undirected dependency graph, sorted
 *      entry points + sorted neighbor visit order.
 *   2. Components larger than `maxClusterSize` are split by directory, then
 *      chunked in `maxClusterSize` slices.
 *   3. Cluster IDs are assigned in final-group order.
 */
export function _clusterFiles(
  files: string[],
  adjacency: Map<string, string[]>,
  maxClusterSize: number = DEFAULT_MAX_CLUSTER_SIZE,
): StructuralMapCluster[] {
  const undirected = new Map<string, Set<string>>();
  for (const f of files) undirected.set(f, new Set());

  for (const [src, targets] of adjacency) {
    for (const tgt of targets) {
      if (undirected.has(tgt)) {
        undirected.get(src)?.add(tgt);
        undirected.get(tgt)?.add(src);
      }
    }
  }

  const visited = new Set<string>();
  const components: string[][] = [];
  const sortedFiles = [...files].sort();

  for (const start of sortedFiles) {
    if (visited.has(start)) continue;
    const component: string[] = [];
    const queue: string[] = [start];
    while (queue.length > 0) {
      const node = queue.shift() as string;
      if (visited.has(node)) continue;
      visited.add(node);
      component.push(node);
      const neighbors = [...(undirected.get(node) ?? new Set<string>())].sort();
      for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) queue.push(neighbor);
      }
    }
    if (component.length > 0) {
      component.sort();
      components.push(component);
    }
  }

  // Split oversized components by directory subtree, then fixed-size chunks.
  const finalGroups: string[][] = [];
  for (const comp of components) {
    if (comp.length <= maxClusterSize) {
      finalGroups.push(comp);
      continue;
    }
    const dirGroups = new Map<string, string[]>();
    for (const f of comp) {
      const d = dirname(f) || '.';
      const bucket = dirGroups.get(d);
      if (bucket) bucket.push(f);
      else dirGroups.set(d, [f]);
    }
    for (const d of [...dirGroups.keys()].sort()) {
      const group = dirGroups.get(d) ?? [];
      for (let i = 0; i < group.length; i += maxClusterSize) {
        finalGroups.push(group.slice(i, i + maxClusterSize));
      }
    }
  }

  const clusters: StructuralMapCluster[] = [];
  for (let idx = 0; idx < finalGroups.length; idx++) {
    const group = finalGroups[idx] as string[];
    const clusterId = idx;

    let name: string;
    if (group.length === 1) {
      const only = group[0] as string;
      const dirPart = dirname(only);
      name = dirPart && dirPart !== '.' ? dirPart : stripExt(only);
    } else {
      name = group.length > 0 ? commonPath(group) : 'root';
    }
    if (!name || name === '.') name = 'root';

    const clusterSet = new Set(group);
    let internalEdges = 0;
    let externalEdges = 0;
    for (const f of group) {
      for (const tgt of adjacency.get(f) ?? []) {
        if (clusterSet.has(tgt)) internalEdges++;
        else externalEdges++;
      }
    }

    clusters.push({
      id: clusterId,
      name,
      files: group,
      internal_edges: internalEdges,
      external_edges: externalEdges,
    });
  }

  return clusters;
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

/**
 * Generate a deterministic PCD structural map for `projectRoot`.
 *
 * If `scope` is provided, only those project-relative paths are analyzed
 * (used by scoped steward-review). Otherwise the shared walker enumerates
 * files honoring `.claude/.prove.json` excludes + size cap. The resulting
 * map is written to `.prove/steward/pcd/structural-map.json` and returned.
 */
export function generateStructuralMap(projectRoot: string, scope?: string[]): StructuralMap {
  const absRoot = resolve(projectRoot);

  let files: string[];
  if (scope !== undefined) {
    files = [...scope];
  } else {
    const config = loadToolConfig(
      absRoot,
      'pcd',
      { excludes: [] as string[], max_file_size: 102400 as number },
      { require: false },
    );
    files = walkProject(absRoot, {
      excludes: config.excludes as string[],
      maxFileSize: config.max_file_size as number,
    });
  }

  const { adjacency } = _buildDependencyGraph(files, absRoot);
  const clusters = _clusterFiles(files, adjacency);

  const fileToCluster = new Map<string, number>();
  for (const cluster of clusters) {
    for (const f of cluster.files) fileToCluster.set(f, cluster.id);
  }

  // Build reverse adjacency (imported_by), deduped + sorted per target.
  const importedBy = new Map<string, string[]>();
  for (const f of files) importedBy.set(f, []);
  for (const [src, targets] of adjacency) {
    for (const tgt of targets) {
      const bucket = importedBy.get(tgt);
      if (bucket) bucket.push(src);
    }
  }
  for (const [key, list] of importedBy) {
    importedBy.set(key, [...new Set(list)].sort());
  }

  const cachePath = join(absRoot, CACHE_PATH);
  const cafiCache = loadCache(cachePath);
  const cafiFiles = cafiCache.files as Record<string, { description?: string } | undefined>;

  const modules: StructuralMapModule[] = [];
  const dependencyEdges: StructuralMapEdge[] = [];
  const languageCounts: Record<string, number> = {};
  let totalLines = 0;

  const sortedFiles = [...files].sort();
  for (const relPath of sortedFiles) {
    const lines = _countLines(join(absRoot, relPath));
    totalLines += lines;

    const language = detectLanguage(relPath) ?? 'unknown';
    languageCounts[language] = (languageCounts[language] ?? 0) + 1;

    const cafiEntry = cafiFiles[relPath];
    const rawDesc = cafiEntry?.description;
    const cafiDesc = rawDesc !== undefined && rawDesc !== '' ? rawDesc : null;

    const importsFrom = [...new Set(adjacency.get(relPath) ?? [])].sort();
    const module: StructuralMapModule = {
      path: relPath,
      lines,
      language,
      exports: [],
      imports_from: importsFrom,
      imported_by: importedBy.get(relPath) ?? [],
      cluster_id: fileToCluster.get(relPath) ?? 0,
    };
    if (cafiDesc !== null) module.cafi_description = cafiDesc;
    modules.push(module);

    for (const tgt of adjacency.get(relPath) ?? []) {
      dependencyEdges.push({ from: relPath, to: tgt, type: 'internal' });
    }
  }

  const structuralMap: StructuralMap = {
    version: 1,
    timestamp: new Date().toISOString().replace('Z', '+00:00'),
    generated_by: 'deterministic',
    summary: {
      total_files: files.length,
      total_lines: totalLines,
      languages: languageCounts,
    },
    modules,
    clusters,
    dependency_edges: dependencyEdges,
  };

  const outputDir = join(absRoot, OUTPUT_DIR);
  mkdirSync(outputDir, { recursive: true });
  const outputPath = join(outputDir, OUTPUT_FILE);
  writeFileSync(outputPath, `${JSON.stringify(structuralMap, null, 2)}\n`, 'utf8');

  return structuralMap;
}

// ---------------------------------------------------------------------------
// Small utilities
// ---------------------------------------------------------------------------

function stripExt(path: string): string {
  const ext = extname(path);
  return ext ? path.slice(0, -ext.length) : path;
}

/**
 * Port of Python's `os.path.commonpath(paths)`:
 *   - Splits each path into segments (POSIX separator), drops empty segs.
 *   - Returns the longest shared segment prefix joined with `/`.
 *   - Assumes inputs are relative paths (walkProject output always is).
 */
function commonPath(paths: string[]): string {
  if (paths.length === 0) return '';
  if (paths.length === 1) return paths[0] ?? '';
  const splits = paths.map((p) => p.split('/').filter((s) => s.length > 0));
  const first = splits[0] ?? [];
  let prefixLen = first.length;
  for (let i = 1; i < splits.length; i++) {
    const next = splits[i] ?? [];
    const minLen = Math.min(prefixLen, next.length);
    let j = 0;
    while (j < minLen && first[j] === next[j]) j++;
    prefixLen = j;
    if (prefixLen === 0) break;
  }
  return first.slice(0, prefixLen).join('/');
}

// Exported only for tests that need to inspect the internal shape.
export { commonPath as _commonPath };
